LeagueToolkit/Helpers/Structures/R3DSphere.cs

```diff
@@ -54,5 +57,93 @@ public void Write(BinaryWriter bw)
             bw.WriteVector3(this.Position);
             bw.Write(this.Radius);
         }
+
+        public bool Contains(Vector3 point)
+        {
+            const double tolerance = 0.001; // need to give a bit of leniency here
+            return Math.Pow(Position.X - point.X, 2) + Math.Pow(Position.Y - point.Y, 2) + Math.Pow(Position.Z - point.Z, 2) <=
+                   Math.Pow(Radius, 2) * (1 + tolerance) + tolerance;
+        }
+
+        public static R3DSphere CalculateBoundingSphere(IEnumerable<Vector3> points)
+        {
+            Random random = new();
+            Vector3[] shuffledPoints = points.OrderBy(_ => random.Next()).ToArray();
+
+            return Welzl(new ArraySegment<Vector3>(shuffledPoints), new List<Vector3>(4));
+        }
+
+        private static R3DSphere Welzl(ArraySegment<Vector3> points, List<Vector3> boundaryPoints)
+        {
+            // see https://en.wikipedia.org/wiki/Smallest-circle_problem#Welzl's_algorithm
+            while (true)
+            {
+                if (points.Count == 0 || boundaryPoints.Count == 4)
+                {
+                    return BoundingSphereFromPoints(boundaryPoints);
+                }
```

Comment:
I don't think that actually works because both `points.Count == 0` and `boundaryPoints.Count == 4` can become true inside the while loop. However in testing I did hit the recursion limit, so I'll have to rewrite the algorithm in an iterative way anyway.

Reply:
Yeah it's written quite strangely currently, although the nature of the algorithm is recursive.
I also had an idea to rewrite it and flatten the recursion as well.